File: backend/app/schemas/graph.py
from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.footprints import Point2D

GraphVariant = Literal["ifc", "geometry", "topologic"]

EdgeMethod = Literal[
    "ifc_rel_space_boundary",
    "same_storey_fallback",
    "vertical_storey_link",
    "geom_door_space",
    "geom_stair_space",
    "geom_opening_space",
    "topologicpy_adjacency",
]

IFC_BASELINE_METHODS: frozenset[str] = frozenset(
    {
        "ifc_rel_space_boundary",
    }
)


class GraphNode(BaseModel):
    id: str
    kind: Literal["space", "door", "stair", "lift"]
    global_id: str
    name: str = ""
    storey_global_id: str | None = None
    """
    True when this space geometrically contains other same-storey spaces
    (candidate to remove or reduce to residual corridor). Geometry variant only;
    nodes are flagged, not removed yet.
    """
    nested_parent: bool = False


class GraphEdge(BaseModel):
    id: str
    kind: Literal["space_door", "vertical", "space_space"]
    source: str
    target: str
    global_id: str | None = None
    method: EdgeMethod
    bidirectional: bool = True
    """True when the edge was not authored via IfcRelSpaceBoundary."""
    inferred: bool = False
    """
    Plan XY for geometric path on space↔space heals: centre of the walkable
    clear span along the shared frontage (used like a door portal).
    """
    portal: Point2D | None = None


class ConnectivityGraph(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    model_id: str
    variant: GraphVariant = "ifc"
    built_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)

File: frontend/src/types/graph.ts
export type GraphNodeKind = "space" | "door" | "stair" | "lift";
export type GraphEdgeKind = "space_door" | "vertical" | "space_space";
export type GraphVariant = "ifc" | "geometry" | "topologic";

export interface GraphNode {
  id: string;
  kind: GraphNodeKind;
  global_id: string;
  name: string;
  storey_global_id: string | null;
  /**
   * Geometry variant: space contains other same-storey spaces
   * (candidate to remove or reduce to residual). Flagged only — not removed yet.
   */
  nested_parent?: boolean;
  /** Optional short code for labels (FE demo). */
  code?: string;
  /** Optional category for colouring (FE demo). */
  category?: string;
}

export interface GraphEdge {
  id: string;
  kind: GraphEdgeKind;
  source: string;
  target: string;
  global_id?: string | null;
  method:
    | "ifc_rel_space_boundary"
    | "same_storey_fallback"
    | "vertical_storey_link"
    | "geom_door_space"
    | "geom_stair_space"
    | "topologicpy_adjacency";
  bidirectional?: boolean;
  /** True when not authored via IfcRelSpaceBoundary — draw green in the viewer. */
  inferred?: boolean;
}

export interface ConnectivityGraph {
  schema_version: "1.0";
  model_id: string;
  variant?: GraphVariant;
  built_at?: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface RouteComputeRequest {
  origin_node_id: string;
  destination_node_id: string;
  blocked_node_ids?: string[];
  blocked_edge_ids?: string[];
  graph?: ConnectivityGraph;
  graph_variant?: GraphVariant;
}

export interface RouteResult {
  found: boolean;
  origin_node_id: string;
  destination_node_id: string;
  node_ids: string[];
  edge_ids: string[];
  hops: number;
  blocked_node_ids: string[];
  blocked_edge_ids: string[];
  message: string;
}

export interface StoreyBand {
  id: string;
  label: string;
  elevation: number;
}
